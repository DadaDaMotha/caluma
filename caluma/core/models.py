import uuid

from django.db import models


class BaseModel(models.Model):
    created_at = models.DateTimeField(auto_now_add=True)
    modified_at = models.DateTimeField(auto_now=True)
    created_by_user = models.CharField(max_length=150, blank=True, null=True)
    created_by_group = models.CharField(
        max_length=150, blank=True, null=True, db_index=True
    )

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """
    Models which use a slug as primary key.

    Defined as Caluma default for configuration so it is possible
    to merge between developer and user configuration.
    """

    slug = models.SlugField(max_length=50, primary_key=True)

    def __str__(self):
        return self.slug

    class Meta:
        abstract = True


class UUIDModel(BaseModel):
    """
    Models which use uuid as primary key.

    Defined as Caluma default
    """

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)

    class Meta:
        abstract = True


class NaturalKeyModel(BaseModel):
    """Models which use a natural key as primary key."""

    id = models.CharField(max_length=255, unique=True, primary_key=True)

    def natural_key(self):  # pragma: no cover
        raise NotImplementedError()

    def save(self, *args, **kwargs):
        self.id = self.natural_key()
        return super().save(*args, **kwargs)

    class Meta:
        abstract = True
