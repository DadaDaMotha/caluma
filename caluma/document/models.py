from django.contrib.postgres.fields import JSONField
from django.db import models

from caluma.models import UUIDModel


class Document(UUIDModel):
    # TODO: add user field once authentication is implemented
    form = models.ForeignKey(
        "form.Form", on_delete=models.DO_NOTHING, related_name="documents"
    )
    meta = JSONField(default={})


class Answer(UUIDModel):
    question = models.ForeignKey(
        "form.Question", on_delete=models.DO_NOTHING, related_name="answers"
    )
    value = JSONField()
    meta = JSONField(default={})
    document = models.ForeignKey(
        Document, on_delete=models.CASCADE, related_name="answers"
    )

    class Meta:
        # a question may only be answerd once per document
        unique_together = ("document", "question")
