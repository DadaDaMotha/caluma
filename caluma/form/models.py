from django.contrib.postgres.fields import JSONField
from django.db import models


class Form(models.Model):
    slug = models.SlugField(max_length=50, primary_key=True)
    name = models.CharField(max_length=255)
    description = models.TextField()
    meta = JSONField()
