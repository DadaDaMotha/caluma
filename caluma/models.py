from django.db import models


class BaseModel(models.Model):
    created = models.DateTimeField(auto_now_add=True)
    modified = models.DateTimeField(auto_now=True)

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """Models which use a slug as primary key."""

    slug = models.SlugField(max_length=50, primary_key=True)

    def __str__(self):
        return self.slug

    class Meta:
        abstract = True
