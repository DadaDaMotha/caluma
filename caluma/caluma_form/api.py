from typing import Any, Optional

from caluma.caluma_form import domain_logic, models
from caluma.caluma_user.models import BaseUser


def save_answer(
    question: models.Question,
    document: models.Document,
    user: Optional[BaseUser] = None,
    value: Optional[Any] = None,
    context: Optional[dict] = None,
    **kwargs
) -> models.Answer:
    """
    Save an answer for given question, document.

    Similar to saveDocumentStringAnswer and the likes, it performes upsert.
    :param value: Must match the question type
    """

    data = {"question": question, "document": document, "value": value}
    data.update(kwargs)

    answer = models.Answer.objects.filter(question=question, document=document).first()

    validated_data = domain_logic.SaveAnswerLogic.pre_save(
        domain_logic.SaveAnswerLogic.validate_for_save(
            data, user, answer=answer, origin=True
        )
    )

    if answer is None:
        answer = domain_logic.SaveAnswerLogic.create(validated_data, user)
    else:
        answer = domain_logic.SaveAnswerLogic.update(validated_data, answer)

    domain_logic.SaveAnswerLogic.post_save(answer)

    return answer
