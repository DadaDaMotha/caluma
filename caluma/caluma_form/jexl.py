from functools import partial

from pyjexl.analysis import ValidatingAnalyzer
from pyjexl.evaluator import Context

from ..caluma_core.jexl import JEXL, ExtractTransformSubjectAnalyzer


class QuestionMissing(Exception):
    pass


class QuestionValidatingAnalyzer(ValidatingAnalyzer):
    def visit_Transform(self, transform):
        if transform.name == "answer" and not isinstance(transform.subject.value, str):
            yield f"{transform.subject.value} is not a valid question slug."

        yield from super().visit_Transform(transform)


class QuestionJexl(JEXL):
    def __init__(self, validation_context=None, **kwargs):
        answer_by_question = validation_context["answers"] if validation_context else {}

        super().__init__(**kwargs)

        context_data = None
        if validation_context:
            context_data = {**validation_context}
            form = validation_context.get("form")
            if form:
                context_data["form"] = form.slug

        self.context = Context(context_data)
        self.answer_by_question = answer_by_question

        self.add_transform("answer", self.answer_transform)
        self.add_transform("mapby", lambda arr, key: [obj[key] for obj in arr])
        self.add_binary_operator(
            "intersects", 20, lambda left, right: any(x in right for x in left)
        )

    def answer_transform(self, question):
        try:
            return self.answer_by_question[question]
        except KeyError:  # pragma: no cover
            raise QuestionMissing(
                f"Question `{question}` could not be found in form {self.context['form']}"
            )

    def validate(self, expression, **kwargs):
        return super().validate(expression, QuestionValidatingAnalyzer)

    def extract_referenced_questions(self, expr):
        transforms = ["answer", "mapby"]
        yield from self.analyze(
            expr, partial(ExtractTransformSubjectAnalyzer, transforms=transforms)
        )

    def _question(self, slug):
        question = self.context["questions"].get(
            slug, self.context["all_questions"].get(slug)
        )
        if not question:
            raise QuestionMissing(
                f"Question `{slug}` could not be found in form {self.context['form']}"
            )
        return question

    def is_hidden(self, question):
        """Return True if the given question is hidden.

        This checks whether the dependency questions are hidden, then
        evaluates the question's is_hidden expression itself.
        """
        # Check visibility of dependencies before actually evaluating the `is_hidden`
        # expression. If all dependencies are hidden,
        # there is no way to evaluate our own visibility, so we default to
        # hidden state as well.

        deps = list(self.extract_referenced_questions(question.is_hidden))

        # all() returns True for the empty set, thus we need to
        # check that we have some deps at all first
        all_deps_hidden = bool(deps) and all(
            self.is_hidden(self._question(dep)) for dep in deps
        )

        return all_deps_hidden or self.evaluate(question.is_hidden)

    def is_required(self, question):
        deps = list(self.extract_referenced_questions(question.is_required))

        if bool(deps) and all(self.is_hidden(self._question(dep)) for dep in deps):
            # all dependent questions are hidden. cannot evaluate,
            # so assume requiredness to be False
            return False

        return self.evaluate(question.is_required)
