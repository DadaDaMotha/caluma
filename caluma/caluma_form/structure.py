"""Hierarchical representation of a document / form."""
import weakref
from functools import singledispatch
from typing import Optional

from .models import Question


def object_local_memoise(meth):
    def new_meth(self, *args, **kwargs):
        if not hasattr(self, "_memoise"):
            self._memoise = {}

        key = str([args, kwargs, meth])
        if key in self._memoise:
            return self._memoise[key]
        ret = meth(self, *args, **kwargs)
        self._memoise[key] = ret
        return ret

    return new_meth


class Element:
    def __init__(self, parent=None):
        self._parent = weakref.ref(parent) if parent else None

    def parent(self):
        return self._parent() if self._parent else None

    def children(self):  # pragma: no cover
        return []


class AnsQuestion(Element):
    def __init__(self, document, form, question, answer=None, parent=None):
        super().__init__(parent)
        self.document = document
        self.form = form
        self.question = question
        self.answer = answer

    @classmethod
    def factory(cls, document, form, question, answer=None, parent=None):
        if question.type == Question.TYPE_FORM:
            return DocForm(
                document, form=question.sub_form, question=question, parent=parent
            )
        elif question.type == Question.TYPE_TABLE:
            return RowAnsQuestion(
                document,
                form=question.row_form,
                question=question,
                answer=answer,
                parent=parent,
            )

        return AnsQuestion(document, form, question, answer, parent=parent)

    def ans_value(self):
        if self.answer is None:
            # no answer object at all - return empty in every case
            return self.question.empty_value()

        if self.answer.value is not None:
            return self.answer.value

        elif self.question.type == Question.TYPE_TABLE:
            return [
                {cell.question.slug: cell.ans_value() for cell in row.children()}
                for row in self.children()
            ]
        elif self.question.type == Question.TYPE_FORM:
            # forms never have values themselves
            return None
            return [
                {cell.question.slug: cell.ans_value() for cell in row.children()}
                for row in self.children()
            ]

    def __repr__(self):
        return f"<AnsQuestion question={self.question.slug}, value={self.ans_value()} hidden=({self.question.is_hidden}) req=({self.question.is_required})>"


class RowAnsQuestion(AnsQuestion):
    @object_local_memoise
    def children(self):
        if not self.answer:
            return []

        return [
            DocForm(
                row_doc,
                self.question.row_form,
                question=self.question,
                parent=self.parent(),
            )
            for row_doc in self.answer.documents.all()
        ]


class DocForm(Element):
    def __init__(self, document, form, question=None, parent=None):
        super().__init__(parent)
        self.document = document
        self.form = form
        self.question = question
        self._ans_questions = None
        self._sub_forms = None

    def ans_value(self):
        return None

    def _build_ans_questions(self):
        if self._ans_questions is None:
            self._ans_questions = {aq.question.slug: aq for aq in self.children()}
        if self._sub_forms is None:
            self._sub_forms = {
                aq.question.slug: aq
                for aq in self.children()
                if aq.question.type == Question.TYPE_FORM
            }

    def get_ans_question(
        self, question_slug: str, allow_check_parent: bool = True
    ) -> Optional[AnsQuestion]:

        self._build_ans_questions()

        ans_question = self._ans_questions.get(question_slug)

        if not ans_question and allow_check_parent:
            ans_question = (
                self.parent().get_ans_question(question_slug) if self.parent() else None
            )
        if ans_question:
            return ans_question

        # OK start looking in subforms / row forms below our level.
        # Since we're looking down, we're disallowing recursing to outer context
        # to avoid recursing back to where we are
        for subform in self._sub_forms.values():
            sub_aq = subform.get_ans_question(question_slug, allow_check_parent=False)
            if sub_aq:
                return sub_aq
        # if we reach this line, we didn't find the question
        return None

    @object_local_memoise
    def children(self):
        answers = {ans.question_id: ans for ans in self.document.answers.all()}
        return [
            AnsQuestion.factory(
                document=self.document,
                form=self.form,
                question=question,
                answer=answers.get(question.slug),
                parent=self,
            )
            for question in self.form.questions.all()
        ]

    def __repr__(self):
        q_slug = self.question.slug if self.question else None
        if q_slug:
            return f"<DocForm fq={q_slug}, doc={self.document.pk} hidden=({self.question.is_hidden}) req=({self.question.is_required})>"

        return f"<DocForm form={self.form.slug}, doc={self.document.pk}>"


def print_document_structure(document):  # pragma: no cover
    """Print a document's structure.

    Intended halfway as an example on how to use the structure
    classes, and a debugging utility.
    """
    ind = {"i": 0}

    @singledispatch
    def visit(vis):
        raise Exception(f"generic visit(): {vis}")

    @visit.register(Element)
    def _(vis):
        print("   " * ind["i"], vis)
        ind["i"] += 1
        for c in vis.children():
            visit(c)
        ind["i"] -= 1

    struc = DocForm(document, document.form)
    visit(struc)
