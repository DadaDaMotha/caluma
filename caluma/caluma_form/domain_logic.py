from django.db import transaction

from caluma.caluma_form import models, validators
from caluma.caluma_user.models import BaseUser
from caluma.utils import update_model


class SaveAnswerLogic:
    @staticmethod
    def validate_for_save(
        data: dict, user: BaseUser, answer: models.Answer = None, origin: bool = False
    ) -> dict:
        question = data["question"]

        if question.type == models.Question.TYPE_TABLE:
            data["documents"] = models.Document.objects.filter(pk__in=data["value"])
            del data["value"]
        elif question.type == models.Question.TYPE_FILE:
            data["file"] = data["value"]
            del data["value"]
        elif question.type == models.Question.TYPE_DATE:
            data["date"] = data["value"]
            del data["value"]

        validators.AnswerValidator().validate(
            **data, user=user, instance=answer, origin=origin
        )

        return data

    @staticmethod
    def pre_save(validated_data: dict) -> dict:
        # TODO emit events
        return validated_data

    @staticmethod
    def post_save(answer: models.Answer) -> models.Answer:
        # TODO emit events
        return answer

    @staticmethod
    @transaction.atomic
    def create(validated_data, user):
        if validated_data["question"].type == models.Question.TYPE_FILE:
            validated_data = __class__.set_file(validated_data)

        if validated_data["question"].type == models.Question.TYPE_TABLE:
            documents = validated_data.pop("documents")

        if user:
            validated_data["created_by_user"] = user
            validated_data["created_by_group"] = user.group

        answer = models.Answer.objects.create(**validated_data)

        if answer.question.type == models.Question.TYPE_TABLE:
            answer.create_answer_documents(documents)

        return answer

    @staticmethod
    @transaction.atomic
    def update(validated_data, answer):
        if answer.question.type == models.Question.TYPE_TABLE:
            documents = validated_data.pop("documents")
            answer.unlink_unused_rows(docs_to_keep=documents)

        if (
            answer.question.type == models.Question.TYPE_FILE
            and answer.file.name is not validated_data["file"]
        ):
            answer.file.delete()
            validated_data = __class__.set_file(validated_data)

        update_model(answer, validated_data)

        if answer.question.type == models.Question.TYPE_TABLE:
            answer.create_answer_documents(documents)

    @staticmethod
    def set_file(validated_data):
        file_name = validated_data.get("file")
        file = models.File.objects.create(name=file_name)
        validated_data["file"] = file
        return validated_data
