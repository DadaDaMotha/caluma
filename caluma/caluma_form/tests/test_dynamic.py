import json

import pytest

from caluma.caluma_form.models import Question


@pytest.mark.parametrize("question__data_source", ["MyDataSourceWithContext"])
@pytest.mark.parametrize(
    "question__type",
    [Question.TYPE_DYNAMIC_CHOICE, Question.TYPE_DYNAMIC_MULTIPLE_CHOICE],
)
@pytest.mark.parametrize("context", [{"foo": "bar"}, None])
def test_dynamic_choice_options_context(
    db, snapshot, question, schema_executor, settings, context
):
    settings.DATA_SOURCE_CLASSES = [
        "caluma.caluma_data_source.tests.data_sources.MyDataSourceWithContext"
    ]

    query = """
        query($question: String!, $context: JSONString) {
            allQuestions(filter: { slugs: [$question] }) {
                edges {
                    node {
                        id
                        ...on DynamicChoiceQuestion {
                            options(context: $context) {
                                edges {
                                    node {
                                        slug
                                        label
                                    }
                                }
                            }
                        }
                        ...on DynamicMultipleChoiceQuestion {
                            options(context: $context) {
                                edges {
                                    node {
                                        slug
                                        label
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    """

    variables = {"question": question.pk}

    if context:
        variables["context"] = json.dumps(context)

    result = schema_executor(query, variable_values=variables)

    assert not result.errors
    snapshot.assert_match(result.data)


@pytest.mark.parametrize("question__data_source", ["MyDataSourceWithContext"])
@pytest.mark.parametrize(
    "question__type,value,context,success",
    [
        (Question.TYPE_DYNAMIC_CHOICE, "option-with-context", {"foo": "bar"}, True),
        (Question.TYPE_DYNAMIC_CHOICE, "option-with-context", None, False),
        (Question.TYPE_DYNAMIC_CHOICE, "option-without-context", None, True),
        (Question.TYPE_DYNAMIC_CHOICE, "option-without-context", {"foo": "bar"}, False),
        (
            Question.TYPE_DYNAMIC_MULTIPLE_CHOICE,
            ["option-with-context"],
            {"foo": "bar"},
            True,
        ),
        (Question.TYPE_DYNAMIC_MULTIPLE_CHOICE, ["option-with-context"], None, False),
        (Question.TYPE_DYNAMIC_MULTIPLE_CHOICE, ["option-without-context"], None, True),
        (
            Question.TYPE_DYNAMIC_MULTIPLE_CHOICE,
            ["option-without-context"],
            {"foo": "bar"},
            False,
        ),
    ],
)
def test_save_dynamic_choice_options_context(
    db, snapshot, question, document, schema_executor, settings, value, context, success
):
    settings.DATA_SOURCE_CLASSES = [
        "caluma.caluma_data_source.tests.data_sources.MyDataSourceWithContext"
    ]

    if question.type == Question.TYPE_DYNAMIC_CHOICE:
        query = """
            mutation($input: SaveDocumentStringAnswerInput!) {
                saveDocumentStringAnswer(input: $input) {
                    answer {
                        ...on StringAnswer {
                            selectedOption {
                                slug
                                label
                            }
                        }
                    }
                }
            }
        """
    elif question.type == Question.TYPE_DYNAMIC_MULTIPLE_CHOICE:
        query = """
            mutation($input: SaveDocumentListAnswerInput!) {
                saveDocumentListAnswer(input: $input) {
                    answer {
                        ...on ListAnswer {
                            selectedOptions {
                                edges {
                                    node {
                                        slug
                                        label
                                    }
                                }
                            }
                        }
                    }
                }
            }
        """

    variables = {
        "input": {
            "question": question.pk,
            "document": str(document.pk),
            "value": value,
            "dataSourceContext": json.dumps(context),
        }
    }

    result = schema_executor(query, variable_values=variables)

    assert bool(result.errors) != success
    snapshot.assert_match(result.data)
