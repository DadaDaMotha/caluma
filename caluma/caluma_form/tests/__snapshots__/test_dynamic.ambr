# name: test_dynamic_choice_options_context[None-dynamic_choice-MyDataSourceWithContext]
  dict({
    'allQuestions': dict({
      'edges': list([
        dict({
          'node': dict({
            'id': 'RHluYW1pY0Nob2ljZVF1ZXN0aW9uOmVudmlyb25tZW50YWwtdGVu',
            'options': dict({
              'edges': list([
                dict({
                  'node': dict({
                    'label': 'q: environmental-ten',
                    'slug': 'option-without-context',
                  }),
                }),
              ]),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_dynamic_choice_options_context[None-dynamic_multiple_choice-MyDataSourceWithContext]
  dict({
    'allQuestions': dict({
      'edges': list([
        dict({
          'node': dict({
            'id': 'RHluYW1pY011bHRpcGxlQ2hvaWNlUXVlc3Rpb246ZW52aXJvbm1lbnRhbC10ZW4=',
            'options': dict({
              'edges': list([
                dict({
                  'node': dict({
                    'label': 'q: environmental-ten',
                    'slug': 'option-without-context',
                  }),
                }),
              ]),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_dynamic_choice_options_context[context0-dynamic_choice-MyDataSourceWithContext]
  dict({
    'allQuestions': dict({
      'edges': list([
        dict({
          'node': dict({
            'id': 'RHluYW1pY0Nob2ljZVF1ZXN0aW9uOmVudmlyb25tZW50YWwtdGVu',
            'options': dict({
              'edges': list([
                dict({
                  'node': dict({
                    'label': 'q: environmental-ten foo: bar',
                    'slug': 'option-with-context',
                  }),
                }),
              ]),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_dynamic_choice_options_context[context0-dynamic_multiple_choice-MyDataSourceWithContext]
  dict({
    'allQuestions': dict({
      'edges': list([
        dict({
          'node': dict({
            'id': 'RHluYW1pY011bHRpcGxlQ2hvaWNlUXVlc3Rpb246ZW52aXJvbm1lbnRhbC10ZW4=',
            'options': dict({
              'edges': list([
                dict({
                  'node': dict({
                    'label': 'q: environmental-ten foo: bar',
                    'slug': 'option-with-context',
                  }),
                }),
              ]),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_choice-option-with-context-None-False-MyDataSourceWithContext]
  dict({
    'saveDocumentStringAnswer': None,
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_choice-option-with-context-context0-True-MyDataSourceWithContext]
  dict({
    'saveDocumentStringAnswer': dict({
      'answer': dict({
        'selectedOption': dict({
          'label': 'q: environmental-ten foo: bar',
          'slug': 'option-with-context',
        }),
      }),
    }),
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_choice-option-without-context-None-True-MyDataSourceWithContext]
  dict({
    'saveDocumentStringAnswer': dict({
      'answer': dict({
        'selectedOption': dict({
          'label': 'q: environmental-ten',
          'slug': 'option-without-context',
        }),
      }),
    }),
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_choice-option-without-context-context3-False-MyDataSourceWithContext]
  dict({
    'saveDocumentStringAnswer': None,
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_multiple_choice-value4-context4-True-MyDataSourceWithContext]
  dict({
    'saveDocumentListAnswer': dict({
      'answer': dict({
        'selectedOptions': dict({
          'edges': list([
            dict({
              'node': dict({
                'label': 'q: environmental-ten foo: bar',
                'slug': 'option-with-context',
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_multiple_choice-value5-None-False-MyDataSourceWithContext]
  dict({
    'saveDocumentListAnswer': None,
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_multiple_choice-value6-None-True-MyDataSourceWithContext]
  dict({
    'saveDocumentListAnswer': dict({
      'answer': dict({
        'selectedOptions': dict({
          'edges': list([
            dict({
              'node': dict({
                'label': 'q: environmental-ten',
                'slug': 'option-without-context',
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_save_dynamic_choice_options_context[dynamic_multiple_choice-value7-context7-False-MyDataSourceWithContext]
  dict({
    'saveDocumentListAnswer': None,
  })
# ---
